python/IECoreMaya/FnSceneShape.py

```diff
@@ -409,74 +409,192 @@ def convertAllToGeometry( self ) :
+
+	@staticmethod
+	def __findOrCreateShape( transformNode, shapeName, shapeType ):
+
+		pathToShape = transformNode + "|" + shapeName
+		fnShape = FnSceneShape.__getFnShape( pathToShape )
+
+		if fnShape and maya.cmds.nodeType( pathToShape ) != shapeType :
 			# Rename existing shape
 			newName = shapeName + "_orig"
-			maya.cmds.rename( shape, newName )
-			IECore.msg( IECore.Msg.Level.Warning, "FnSceneShape.convertObjectToGeometry", "Renaming incompatible shape %s to %s." % shape, newName )
+			maya.cmds.rename( pathToShape, newName )
+			IECore.msg( IECore.Msg.Level.Warning, "FnSceneShape.__findOrCreateShape", "Renaming incompatible shape %s to %s." % pathToShape, newName )
 			fnShape = None
 
 		if not fnShape :
 			dagMod = maya.OpenMaya.MDagModifier()
-			shapeNode = dagMod.createNode( type, IECoreMaya.StringUtil.dependencyNodeFromString( transformNode ) )
+			shapeNode = dagMod.createNode( shapeType, IECoreMaya.StringUtil.dependencyNodeFromString( transformNode ) )
 			dagMod.renameNode( shapeNode, shapeName )
 			dagMod.doIt()
 			
 			fnShape = maya.OpenMaya.MFnDagNode( shapeNode )
 
-			if type == "mesh":
-				maya.cmds.sets(shape, add="initialShadingGroup" )
+			if shapeType == "mesh":
+				maya.cmds.sets( pathToShape, add="initialShadingGroup" )
 
-		plug = fnShape.findPlug( plug )
-		if plug.isLocked() :
-			return
-		
-		connections = maya.OpenMaya.MPlugArray()
-		if plug.isConnected() :
-			plug.connectedTo( connections, True, False )
-		
-		if not connections.length() :
+		return fnShape
+
+	def __findOrCreateShapes( self, transformNode ):
+
+		shapeType, plugStr = self.__mayaCompatibleShapeAndPlug()
+		if not (shapeType and plugStr):
+			raise Exception, "Scene interface at %s cannot be converted to Maya geometry." % self.sceneInterface().pathAsString()
+
+		shapeName = IECoreMaya.FnDagNode.defaultShapeName( transformNode )
+
+		if shapeType == "nurbsCurve":
+			curvesPrimitive = self.sceneInterface().readObject( 0.0 )
+			numShapes = curvesPrimitive.numCurves()
+			for shapeId in range( numShapes ):
+				FnSceneShape.__findOrCreateShape( transformNode, shapeName + str( shapeId ), shapeType )
+
+		else:
+			FnSceneShape.__findOrCreateShape( transformNode, shapeName, shapeType )
+
+	def __connectShape( self, pathToShape, plugStr, arrayIndex ):
+
+			fnShape = FnSceneShape.__getFnShape( pathToShape )
+
+			plug = fnShape.findPlug( plugStr )
+			if plug.isLocked() :
+				return
+
+			connections = maya.OpenMaya.MPlugArray()
+			if plug.isConnected() :
+				plug.connectedTo( connections, True, False )
+			if connections.length():
+				return
+
+			# Connect this node to the shape.
 			dgMod = maya.OpenMaya.MDGModifier()
-			dgMod.connect( self.findPlug( "outObjects" ).elementByLogicalIndex( self.__queryIndexForPath( "/" ) ), plug )
+			dgMod.connect( self.findPlug( "outObjects" ).elementByLogicalIndex( arrayIndex ), plug )
 			dgMod.doIt()
-			
-			if type == "mesh":
+
+			# todo: Should it be here?
```

Comment:
can this part be moved up into `__findOrCreateShape`? The interpolation attr on the mesh is considered part of the topology, so makes more sense in the "create" stage than the "connect" stage I guess.